File: zikzak_inappwebview_android/android/src/main/java/wtf/zikzak/zikzak_inappwebview_android/headless_in_app_webview/HeadlessInAppWebViewManager.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

package wtf.zikzak.zikzak_inappwebview_android.headless_in_app_webview;

import android.content.Context;

import androidx.annotation.Nullable;
import androidx.annotation.NonNull;

import wtf.zikzak.zikzak_inappwebview_android.InAppWebViewFlutterPlugin;
import wtf.zikzak.zikzak_inappwebview_android.types.ChannelDelegateImpl;
import wtf.zikzak.zikzak_inappwebview_android.webview.in_app_webview.FlutterWebView;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.Result;

public class HeadlessInAppWebViewManager extends ChannelDelegateImpl {
  protected static final String LOG_TAG = "HeadlessInAppWebViewManager";
  public static final String METHOD_CHANNEL_NAME = "wtf.zikzak/flutter_headless_inappwebview";

  public final Map<String, HeadlessInAppWebView> webViews = new HashMap<>();
  @Nullable
  public InAppWebViewFlutterPlugin plugin;

  public HeadlessInAppWebViewManager(final InAppWebViewFlutterPlugin plugin) {
    super(new MethodChannel(plugin.messenger, METHOD_CHANNEL_NAME));
    this.plugin = plugin;
  }

  @Override
  public void onMethodCall(final MethodCall call, @NonNull final Result result) {
    final String id = (String) call.argument("id");

    switch (call.method) {
      case "run":
        {
          HashMap<String, Object> params = (HashMap<String, Object>) call.argument("params");
          run(id, params, () -> result.success(true));
        }
        break;
      default:
        result.notImplemented();
    }
  }

  public void run(String id, HashMap<String, Object> params, Runnable completion) {
    if (plugin == null || (plugin.activity == null && plugin.applicationContext == null)) {
      completion.run();
      return;
    }
    Context context = plugin.activity;
    if (context == null) {
      context = plugin.applicationContext;
    }
    FlutterWebView flutterWebView = new FlutterWebView(plugin, context, id, params);
    HeadlessInAppWebView headlessInAppWebView = new HeadlessInAppWebView(plugin, id, flutterWebView);
    webViews.put(id, headlessInAppWebView);

    headlessInAppWebView.prepare(params);
    headlessInAppWebView.onWebViewCreated();

    // Readiness gate: completes run() only after the initial navigation
    // reaches a terminal state (onPageFinished or a main-frame error). This
    // guarantees the webview / renderer is up before a consumer issues its
    // first real loadUrl. Signal-driven on purpose: no timeout constant — a
    // missing terminal event is a wiring bug that must surface loudly, not be
    // masked by a magic number.
    if (flutterWebView.webView == null) {
      completion.run();
      return;
    }
    // Nothing to wait for if no initial load was requested.
    if (params.get("initialUrlRequest") == null
        && params.get("initialFile") == null
        && params.get("initialData") == null) {
      completion.run();
      return;
    }
    final boolean[] gateFired = {false};
    flutterWebView.webView.firstNavigationCompleted = () -> {
      if (gateFired[0]) return;
      gateFired[0] = true;
      flutterWebView.webView.firstNavigationCompleted = null;
      completion.run();
    };

    flutterWebView.makeInitialLoad(params);
  }

  @Override
  public void dispose() {
    super.dispose();
    Collection<HeadlessInAppWebView> headlessInAppWebViews = webViews.values();
    for (HeadlessInAppWebView headlessInAppWebView : headlessInAppWebViews) {
      if (headlessInAppWebView != null) {
        headlessInAppWebView.dispose();
      }
    }
    webViews.clear();
    plugin = null;
  }
}

File: zikzak_inappwebview_ios/ios/zikzak_inappwebview_ios/Sources/zikzak_inappwebview_ios/HeadlessInAppWebView/HeadlessInAppWebViewManager.swift
//
//  HeadlessInAppWebViewManager.swift
//  zikzak_inappwebview
//
//  Created by ARRRRNY on 10/05/2020.
//

import UIKit

import Flutter
import UIKit
import WebKit
import UIKit
import AVFoundation

public class HeadlessInAppWebViewManager: ChannelDelegate {
    static let METHOD_CHANNEL_NAME = "wtf.zikzak/flutter_headless_inappwebview"
    var plugin: SwiftFlutterPlugin?
    var webViews: [String: HeadlessInAppWebView?] = [:]

    init(plugin: SwiftFlutterPlugin) {
        super.init(channel: FlutterMethodChannel(name: HeadlessInAppWebViewManager.METHOD_CHANNEL_NAME, binaryMessenger: plugin.registrar!.messenger()))
        self.plugin = plugin
    }

    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? NSDictionary
        let id: String = arguments!["id"] as! String

        switch call.method {
            case "run":
                let params = arguments!["params"] as! [String: Any?]
                run(id: id, params: params) {
                    result(true)
                }
                break
            default:
                result(FlutterMethodNotImplemented)
                break
        }
    }

    public func run(id: String, params: [String: Any?], completion: @escaping () -> Void) {
        guard let plugin = plugin else {
            completion()
            return
        }
        let flutterWebView = FlutterWebViewController(plugin: plugin,
            withFrame: CGRect.zero,
            viewIdentifier: id,
            params: params as NSDictionary)
        let headlessInAppWebView = HeadlessInAppWebView(plugin: plugin, id: id, flutterWebView: flutterWebView)
        webViews[id] = headlessInAppWebView

        headlessInAppWebView.prepare(params: params as NSDictionary)
        headlessInAppWebView.onWebViewCreated()

        // Readiness gate: WKWebView can silently DROP a navigation issued
        // while its content process is still booting (the initial about:blank
        // load from makeInitialLoad is in flight and the web process is
        // spawning). A consumer calling loadUrl immediately after run()
        // would see the first real navigation never start — no didStart,
        // no didFinish — and would only discover it via its own timeout.
        // So run() completes only after the initial navigation reaches a
        // terminal state (didFinish or didFail). Signal-driven on purpose:
        // no timeout constant — if no terminal event ever fires, that is a
        // wiring bug that must surface loudly, not be masked by a magic number.
        let webView = flutterWebView.webView()
        var gateFired = false
        webView?.firstNavigationCompleted = { [weak webView] in
            guard !gateFired else { return }
            gateFired = true
            webView?.firstNavigationCompleted = nil
            completion()
        }

        flutterWebView.makeInitialLoad(params: params as NSDictionary)
    }

    public override func dispose() {
        super.dispose()
        let headlessWebViews = webViews.values
        headlessWebViews.forEach { (headlessWebView: HeadlessInAppWebView?) in
            headlessWebView?.dispose()
        }
        webViews.removeAll()
        plugin = nil
    }

    deinit {
        dispose()
    }
}
